fenix/app/src/main/java/org/mozilla/fenix/settings/quicksettings/protections/cookiebanners/CookieBannersStorageExt.kt

```diff
@@ -0,0 +1,56 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.settings.quicksettings.protections.cookiebanners
+
+import android.content.Context
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.concept.engine.cookiehandling.CookieBannersStorage
+import org.mozilla.fenix.ext.settings
+import org.mozilla.fenix.trackingprotection.CookieBannerUIMode
+
+/**
+ * Get the current status of cookie banner ui mode.
+ */
+suspend fun CookieBannersStorage.getCookieBannerUIMode(
+    context: Context,
+    tab: SessionState,
+    cookieBannerModeResult: (cookieBannerUIMode: CookieBannerUIMode) -> Any,
+) {
+    if (context.settings().shouldUseCookieBanner) {
+        val hasException = withContext(Dispatchers.IO) {
+            hasException(
+                tab.content.url,
+                tab.content.private,
+            )
+        }
+        if (hasException) {
+            cookieBannerModeResult.invoke(CookieBannerUIMode.DISABLE)
+        } else {
+            showUnsupportedSiteIfNeeded(tab, cookieBannerModeResult)
+        }
+    } else {
+        cookieBannerModeResult.invoke(CookieBannerUIMode.HIDE)
+    }
+}
+
+private fun showUnsupportedSiteIfNeeded(
+    tab: SessionState,
+    cookieBannerModeResult: (cookieBannerUIMode: CookieBannerUIMode) -> Any,
+) {
+    tab.engineState.engineSession?.hasCookieBannerRuleForSession(
+        onResult = { result ->
+            if (result) {
+                cookieBannerModeResult.invoke(CookieBannerUIMode.ENABLE)
+            } else {
+                cookieBannerModeResult.invoke(CookieBannerUIMode.SITE_NOT_SUPPORTED)
+            }
+        },
+        onException = {
+            cookieBannerModeResult.invoke(CookieBannerUIMode.HIDE)
+        },
+    )
+}
```

Comment:
Done

Reply:
tab.isCookieBannerSupported() should be   withContext(Dispatchers.Main) {tab.isCookieBannerSupported()} 
because I get a crash .
Uncaught exception handled: 
                                                                                                    java.lang.IllegalThreadStateException: Must have a Handler
                                                                                                    	at org.mozilla.geckoview.GeckoResult.then(GeckoResult.java:567)
                                                                                                    	at mozilla.components.browser.engine.gecko.GeckoEngineSession.hasCookieBannerRuleForSession(GeckoEngineSession.kt:450)
                                                                                                    	at org.mozilla.fenix.settings.quicksettings.protections.cookiebanners.CookieBannersStorageExtKt.isCookieBannerSupported(CookieBannersStorageExt.kt:40)
                                                                                                    	at org.mozilla.fenix.settings.quicksettings.protections.cookiebanners.CookieBannersStorageExtKt.getCookieBannerUIMode(CookieBannersStorageExt.kt:31)
                                                                                                    	at org.mozilla.fenix.settings.quicksettings.protections.cookiebanners.CookieBannersStorageExtKt$getCookieBannerUIMode$1.invokeSuspend(Unknown Source:13)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)